File: test/com/barrybecker4/simulation/lsystem/model/expression/LExpressionParserTest.java
// Copyright by Barry G. Becker, 2013. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.lsystem.model.expression;

import com.barrybecker4.common.expression.TreeNode;
import junit.framework.TestCase;

/**
 * @author Barry Becker
 */
public class LExpressionParserTest extends TestCase {

    /** instance under test */
    private LExpressionParser parser;

    /** used to verify parsed tree */
    private LTreeSerializer serializer;

    /**
     * common initialization for all go test cases.
     */
    @Override
    protected void setUp() throws Exception {
        super.setUp();

        parser = new LExpressionParser();
        serializer = new LTreeSerializer();
    }


    public void testFOnlyExp() {
        verifyParse("F");
    }

    public void testSimpleExp() {
        verifyParse("F(-F)");
    }

    public void testExpWithSpaces() {
        verifyParse("F (- F)", "F(-F)");
    }

    public void testTwoLevelNestedExp() {
        verifyParse("FF(F(-F))(++F)");
    }

    /**
     * @param exp the expression to parse
     */
    private void verifyParse(String exp) {
        verifyParse(exp, exp);
    }


    /**
     * @param exp the expression to parse
     */
    private void verifyParse(String exp, String expSerializedStr) {
        TreeNode root = parser.parse(exp);
        String serialized = serializer.serialize(root);
        assertEquals("Unexpected parsed expression tree.",
                expSerializedStr, serialized);
    }

}

File: scala-test/com/barrybecker4/simulation/lsystem/model/expression/LExpressionParserTest.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.lsystem.model.expression

import org.scalatest.FunSuite


/**
  * @author Barry Becker
  */
class LExpressionParserTest extends FunSuite {

  /** instance under test */
  private var parser = new LExpressionParser
  /** used to verify parsed tree */
  private var serializer = new LTreeSerializer

  test("FOnlyExp") { verifyParse("F") }
  test("SimpleExp") { verifyParse("F(-F)") }
  test("ExpWithSpaces") { verifyParse("F (- F)", "F(-F)") }
  test("TwoLevelNestedExp") { verifyParse("FF(F(-F))(++F)") }

  /**  @param exp the expression to parse */
  private def verifyParse(exp: String) { verifyParse(exp, exp) }

  private def verifyParse(exp: String, expSerializedStr: String) {
    val root = parser.parse(exp)
    val serialized = serializer.serialize(root)
    assertResult(expSerializedStr) { serialized }
  }
}
